### ISSUE TITLE
A version control branching model for scientific software development

### ISSUE THREAD

**Issue by markcoletti** [betterscientificsoftware/bssw.io#1672]
Version control systems support branching, where developers can isolate their work-in-progress from their peers.  Typically when a new feature or bug fix is started, a developer will create a branch, perform the work, and then merge the results into the base branch from which their branch was made.  Git Flow is a branching model that has standard branches for production releases, called the main branch, and a developer branch from which new features and bug fixes are made. However, developers will create a feature branch from the developer branch to implement new functionality or code repairs; work on a feature branch is merged into the develop branch when done.  There are other branch types for making emergency hot fixes and for performing releases.

However, git flow is fine for enterprise level software development, but is overkill for research software development.  Here we share an alternative and simpler branching model that's suitable for scientific software development that has been effective at the Oak Ridge National Laboratory.

Mark Coletti, PhD
colettima@ornl.gov

**Comment by bartlettroscoe** [comment]
Is this workflow different than a combination of a topic-branch workflow (i.e. [GitHub Flow](https://docs.github.com/en/get-started/quickstart/github-flow)) and a simple release branch workflow (i.e. 'maint' in [gitworkflows](https://git-scm.com/docs/gitworkflows))?

**Comment by markcoletti** [comment]
Essentially, it’s creating feature branches off the main branch; but also, the feature branches have an associated issue on a Kanban board (which both github and gitlab support) and name the branch from that issue.  In an earlier version of this branching model we we’d name the branch something like “issue-3”, but both gitlab and github now support creating new branches from newly created issues, in which case the name will be derived from the issue title.

This recognizes that much scientific software development is done by scientists and not by software engineers, and therefore may not be comfortable with complex branching models. Also, most scientific software is ad hoc, one off, so there is no notion of a release, or hotfixes, etc., therefore a lot of the formal git flow branching model is irrelevant in that context.

Also, the point will be raised that this approach can accommodate managing experiments.  A lot of experiments are done ad hoc, which can lead to all manner of problems.  But leaning on a Kanban board with an associated branching model adds some helpful scientific project discipline.

From: Roscoe A. Bartlett ***@***.***>
Date: Thursday, May 4, 2023 at 2:31 PM
To: betterscientificsoftware/bssw.io ***@***.***>
Cc: Coletti, Mark ***@***.***>, Author ***@***.***>
Subject: [EXTERNAL] Re: [betterscientificsoftware/bssw.io] A version control branching model for scientific software development (Issue #1672)

Is this workflow different than a combination of a topic-branch workflow (i.e. GitHub Flow<https://docs.github.com/en/get-started/quickstart/github-flow>) and a simple release branch workflow (i.e. 'maint' in gitworkflows<https://git-scm.com/docs/gitworkflows>)?

**Comment by markcmiller86** [comment]
> Also, most scientific software is ad hoc, one off, so there is no notion of a release, or hotfixes, etc.,

I'd be careful and qualify this a bit. Maybe thats *most* of what you encounter in the neck of the woods you are currently traveling. I've traveled in such woods myself.

But, in most of the parts of the HPC/CSE community I currently frequent, I see releases, I see hotfixes and I see thorough design.

**Comment by markcoletti** [comment]
Duly noted, though I’ve been involved in many research software engineering projects – some do follow the git flow model, particularly open-source scientific software.  But a lot of the projects I’m involved with use bespoke code that’s never seen again.  However, yes, we’re sharing anecdotal experiences, and I’ve got to be mindful that “the plural of anecdote isn’t data.”

From: markcmiller86 ***@***.***>
Date: Thursday, May 4, 2023 at 3:26 PM
To: betterscientificsoftware/bssw.io ***@***.***>
Cc: Coletti, Mark ***@***.***>, Author ***@***.***>
Subject: [EXTERNAL] Re: [betterscientificsoftware/bssw.io] A version control branching model for scientific software development (Issue #1672)

Also, most scientific software is ad hoc, one off, so there is no notion of a release, or hotfixes, etc.,

I'd be careful and qualify this a bit. Maybe thats most of what you encounter in the neck of the woods you are currently traveling. I've traveled in such woods myself.

But, in most of the parts of the HPC/CSE community I currently frequent, I see releases, I see hotfixes and I see thorough design.

**Comment by markcoletti** [comment]
So, do I have the go ahead to write and submit this article?

**Comment by bartlettroscoe** [comment]
> So, do I have the go ahead to write and submit this article?

@markcoletti, yes, please submit a draft for the article as a pull request submitting a new `*.md` file.  I think a blog article is likely the most appropriate.  Details on those types of article are given at:

* https://betterscientificsoftware.github.io/bssw.io/bssw_styling_originalarticles.html

I would suggest that you fork the bssw.io GitHub repo, clone the bssw.io repo locally, checkout a new branch, and copy a recent blog article file like [Articles/Blog/2023-03-Julia_for_BSSw.md](https://github.com/betterscientificsoftware/bssw.io/blob/master/Articles/Blog/2023-03-Julia_for_BSSw.md) with your own article file name.  When create your GitHub PR, it will give you a template with a checklist.  We will then take over from there and communicate with you there through the PR.

**Comment by majidaldo** [comment]
> 

depends on the phase. when things are exploratory, as in data science, ya the git flow is overkill.
however, over time things firm up and can make use of git flow.

**Comment by majidaldo** [comment]
> > Also, most scientific software is ad hoc, one off, so there is no notion of a release, or hotfixes, etc.,
> 
> I'd be careful and qualify this a bit. Maybe thats _most_ of what you encounter in the neck of the woods you are currently traveling. I've traveled in such woods myself.
> 
> But, in most of the parts of the HPC/CSE community I currently frequent, I see releases, I see hotfixes and I see thorough design.

depends on the phase. when things are exploratory, as in data science, ya the git flow is overkill.
however, over time things firm up and can make use of git flow.

**Comment by markcmiller86** [comment]
Just pinging this issue to inquire as to status or progress. Please let us know.

**Comment by markcoletti** [comment]
> Just pinging this issue to inquire as to status or progress. Please let us know.

Thanks for the ping.  Been super busy, but I do intend on finishing this article.

It's my birthday today, so ... probably not today!  :D

**Comment by markcmiller86** [comment]
🎂

**Comment by bernhold** [comment]
Hi @markcoletti given the time that has elapsed, I'm guessing this has dropped off your radar, so I'm going to close it.  Please feel free to reopen if you still want to write it.  Or submit another idea.  Thanks for your interest in BSSw.io!
